File: verdeciel/Empty.swift

//  Created by Devine Lu Linvega on 2016-04-22.
//  Copyright © 2016 XXIIVV. All rights reserved.

import UIKit
import QuartzCore
import SceneKit
import Foundation

class Empty : SCNNode
{
	var details:String = "unknown"
	
	override init()
	{
		super.init()
	}
	
	func touch(id:Int = 0)
	{
	}
	
	func update()
	{
	}
	
	func empty()
	{
		for node in childNodes {
			node.removeFromParentNode()
		}
	}
	
	func blink()
	{
		if game.time % 3 == 0 { opacity = 1 }
		else { opacity = 0 }
	}
	
	func show()
	{
		opacity = 1
	}
	
	func hide()
	{
		opacity = 0
	}
	
	func updateChildrenColors(color:UIColor)
	{
		for node in self.childNodes as! [Empty] {
			node.color(color)
			node.updateChildrenColors(color)
		}
	}
	
	func color(color:UIColor)
	{
		if geometry == nil { return }
		if (self is SCNLine) == false { return }
		
		(self as! SCNLine).update(color)
	}
	
	// MARK: Triggers -
	
	func onConnect()
	{
		update()
	}
	
	func onDisconnect()
	{
		update()
	}
	
	func onUploadComplete()
	{
		
	}
	
	func onMissionComplete()
	{
		
	}
	
	// MARK: Extras -
	
	func payload() -> ConsolePayload
	{
		return ConsolePayload(data:[ConsoleData(text: "unknown", details: "unknown")])
	}
	
	required init?(coder aDecoder: NSCoder)
	{
		fatalError("init(coder:) has not been implemented")
	}
}

File: scripts/components/empty.js
//  Created by Devine Lu Linvega.
//  Copyright © 2017 XXIIVV. All rights reserved.

class Empty extends SceneNode {
  constructor() {
    // assertArgs(arguments, 0);
    super();
    this.details = "unknown";
  }

  touch(id = 0) {
    // assertArgs(arguments, 1);
    return false;
  }

  update() {
    // assertArgs(arguments, 0, true);
  }

  empty() {
    // assertArgs(arguments, 0);
    while (this.children.length > 0) {
      this.remove(this.children[0]);
    }
  }

  blink() {
    // assertArgs(arguments, 0);
    if (verreciel.game.time % 3 == 0) {
      this.opacity = 1;
    } else {
      this.opacity = 0;
    }
  }

  show() {
    // assertArgs(arguments, 0);
    this.opacity = 1;
  }

  hide() {
    // assertArgs(arguments, 0);
    this.opacity = 0;
  }

  updateChildrenColors(color) {
    for (let node of this.children) {
      node.updateChildrenColors(color);
    }
  }

  onConnect() {
    // assertArgs(arguments, 0);
    this.update();
  }

  onDisconnect() {
    // assertArgs(arguments, 0);
    this.update();
  }

  onUploadComplete() {
    // assertArgs(arguments, 0);
  }

  onMissionComplete() {
    // assertArgs(arguments, 0);
  }

  payload() {
    // assertArgs(arguments, 0);
    return new ConsolePayload([new ConsoleData("unknown", "unknown")]);
  }
}
